open-company/open-company-api
# Lock down write access to company's org members and provide anonymous public access

https://trello.com/c/nOF9AoKg

This PR BOTH opens up access to anonymous API users w/o a JWToken AND locks down access to only a company's organization. Before we required a JWToken for everything, but any valid JWToken would do for any request. Now it's more complicated, requests are either:

OK for anonymous:
- invalid JWToken **not** allowed
- **no** JWToken allowed
- any valid JWToken allowed

OK only for company org. members:
- invalid JWToken **not** allowed
- **no** JWToken **not** allowed
- valid JWToken for another org **not** allowed
- valid JWToken for company's org allowed

This spreadsheet details the security rules and expected response for every API call in use at this point:
https://docs.google.com/spreadsheets/d/1AvDqPHzDTWUxUdpzw9Hlff4IWbxc2QwHvcwGbVrYoSg/edit#gid=0

This PR implements the rules and responses of that spreadsheet, and provides tests for the same.

To test:
0. Review the code changes
1. Review the tests, compare them to the [spreadsheet](https://docs.google.com/spreadsheets/d/1AvDqPHzDTWUxUdpzw9Hlff4IWbxc2QwHvcwGbVrYoSg/edit#gid=0). Do they adequately test everything there?
2. Run the tests with `lein test!`
3. Reimport all the sample company data (it now stores the organization)
4. Double check any unit tests you want to using the REPL
5. Double check any integration tests you want to using cURL and the client UI
6. For the client UI, **remove your JWToken cookies first**, and it'll be handy to test with the company `read-only`, take a look at the company GET request, it has no HATEOAS links except for 'self' on company and sections? Take a look at the OPTIONS request for `sections/new`, it 401's? Groovy!
7. Merge
8. Deploy to staging/prod
9. Reimport all sample companies staging/prod
10. Sing the Star Spangled Banner a cappella or with ukulele accompaniment (no guitar!)
